src/App/Fossa/Test.hs

```diff
@@ -0,0 +1,174 @@
+
+module App.Fossa.Test
+  ( testMain
+  ) where
+
+import Prologue
+import qualified Prelude as Unsafe
+
+import qualified App.Fossa.Analyze.FossaV1 as Fossa
+import App.Fossa.Analyze.ProjectInference
+import Control.Carrier.Error.Either
+import Control.Concurrent (threadDelay)
+import qualified Control.Concurrent.Async as Async
+import qualified Data.Map as M
+import qualified Data.Text as T
+import Data.Text.IO (hPutStrLn)
+import Effect.Logger
+import Network.HTTP.Req
+import Path.IO
+import System.IO (stderr)
+import System.Exit (exitSuccess, exitFailure)
+
+pollDelaySeconds :: Int
+pollDelaySeconds = 8
+
+testMain
+  :: Url 'Https -- ^ api base url
+  -> Text -- ^ api key
+  -> Severity
+  -> Int -- ^ timeout (seconds)
+  -> Maybe Text -- ^ cli override for name
+  -> Maybe Text -- ^ cli override for revision
+  -> IO ()
+testMain baseurl apiKey logSeverity timeoutSeconds overrideName overrideRevision= do
+  basedir <- getCurrentDir
+
+  void $ timeout timeoutSeconds $ withLogger logSeverity $ do
```

Comment:
This magic timeout is very cool, but unfortunately forces us to directly use IO here. The existing fused-effects machinery doesn't allow us to express timeout.

As soon as it does, the code in Fossa.Main, Fossa.Analyze, and Fossa.Test gets substantially simpler